Add `pulumi query` command

hausdorff opened the pull request:
This is a pretty hefty review. I recommend looking at it commit-by-commit, as I tried to structure it into bite-sized bits with good commit messages.

Things to know about the PR:

* `pulumi query` is effectively "run the current project in 'query mode'." This mode disallows all resource operations, but allows access to the stack through the default provider. You can change the project being executed in "query mode" with `--cwd` or `-C` at the CLI.
* Currently `query` does _not_ support simply "invoking" files -- you still do need a whole project.
* `pulumi up` should be nearly completely unaffected by `pulumi query`. In particular, `query` does not use any of the plan or step execution machinery, and mostly does not use the engine. This makes `query` a "mostly parallel universe" to the update path, so there are minimal conditionals and special cases.
* `query` ships its own `QuerySource`, which is kind of like a very lightweight `EvalSource`. It essentially just manages the execution of the language plugins and the builtin provider. It has no iterator, just a `Wait` method.
* `query` ships `queryResmon`, its own resource monitor that simply errors out for all resource operations, but still provides access to stack snapshots through `Invoke` on the builtin provider. `queryResmon` (purposefully) does _not_ support resource plugin resolution at this point.
* `query` ships a new display mode, alongside diff and progress views. It's designed to let the language plugin look like "just another program", rather than going through the `up` display or the diff display.

**Still working on:**

- [ ] Tests!
- [ ] We still use some of the "helper" structs like `UpdateInfo` and `UpdateOptions`. This isn't the end of the world, but it would be nice to clean it up a bit before shipping. We'll see if that's worth the time.

hausdorff commented:
@swgillespie @ellismg @lukehoban ok, this is close to being in shape. I will be adding tests and buttoning things up, but I'd appreciate early comments anyway.

justinvp commented on cmd/query.go:
```diff
@@ -0,0 +1,122 @@
+	var configArray []string
+
+	// up implementation used when the source of the Pulumi program is in the current working directory.
+	upWorkingDirectory := func(opts backend.UpdateOptions) result.Result {
```
This whole func can be deleted and its content moved inline on line 107

justinvp commented on cmd/query.go:
```diff
@@ -0,0 +1,122 @@
+		}
+	}
+
+	// up implementation used when the source of the Pulumi program is a template name or a URL to a template.
```
This can be deleted.

justinvp commented on cmd/query.go:
```diff
@@ -0,0 +1,122 @@
+	// up implementation used when the source of the Pulumi program is a template name or a URL to a template.
+
+	var cmd = &cobra.Command{
+		Use:   "query [template|url]",
```
We don't support passing templates to `pulumi query`, right?

justinvp commented on cmd/query.go:
```diff
@@ -0,0 +1,122 @@
+		"Use the configuration values in the specified file rather than detecting the file name")
+	cmd.PersistentFlags().StringArrayVarP(
+		&configArray, "config", "c", []string{},
+		"Config to use during the update")
```
This flag was added to provide a way of updating config for Pulumi apps used with `pulumi up <template>` (as the source for such apps would never be available in a place on disk where you could run `pulumi config set`). We can probably just leave it off of `pulumi query` (`pulumi preview` doesn't have it).

justinvp commented on cmd/query.go:
```diff
@@ -0,0 +1,122 @@
+// Copyright 2016-2018, Pulumi Corporation.
```
Nit: 2019

justinvp commented on cmd/query.go:
```diff
@@ -0,0 +1,122 @@
+			"\n" +
+			"The program to run is loaded from the project in the current directory by default. Use the `-C` or\n" +
+			"`--cwd` flag to use a different directory.",
+		Args: cmdutil.MaximumNArgs(1),
```
At which point this can be `cmdutil.NoArgs`.

swgillespie commented on pkg/backend/display/query.go:
```diff
@@ -0,0 +1,120 @@
+	}
+
+	// For stdout messages, trim ONLY the last newline character.
+	if payload.Severity == diag.Info && strings.HasSuffix(payload.Message, "\n") {
```
This could get weird on Windows, where newlines can be `\r\n` and in this case would get shortened to `\r`. Have you tried this on Windows?

hausdorff replied:
Oh, no, that's a good point. We should be using the OS-specific NL.

hausdorff replied:
Actually, wait, Go normalizes newlines so that you only need to know about `\n`, doesn't it?

swgillespie replied:
I dunno, actually.

hausdorff replied:
No, apparently they don't. I'll try it on Windows.

hausdorff replied:
@swgillespie lol turns out we already have a helper function for this: `cmdutil.RemoveTralingNewline`.

swgillespie commented on pkg/engine/query.go:
```diff
@@ -0,0 +1,198 @@
+	// Note that this is purely a best-effort thing. If we can't install missing plugins, just proceed; we'll fail later
+	// with an error message indicating exactly what plugins are missing.
+	if err := ensurePluginsAreInstalled(client, allPlugins); err != nil {
+		logging.V(7).Infof("newUpdateSource(): failed to install missing plugins: %v", err)
```
nit: `newQueryIterator`

hausdorff replied:
Let's call it `newQuerySource`.

swgillespie commented on pkg/resource/deploy/source_query.go:
```diff
@@ -0,0 +1,272 @@
+	label := fmt.Sprintf("QueryResourceMonitor.Invoke(%s)", tok)
+
+	// Dispatch resource output request to default provider.
+	if tok == readStackResourceOutputs {
```
It might be a little easier to read this function if we invert this `if` condition and immediately return if we're not doing the one allowed invoke. That way it's a lot easier to be absolutely sure that we're not accidentally following through the branch without returning and accidentally returning an error.

swgillespie commented on cmd/query.go:
```diff
@@ -0,0 +1,98 @@
+				Type:          display.DisplayQuery,
+			}
+
+			return func(opts backend.UpdateOptions) result.Result {
```
Any reason not to inline this function into the containing function?

swgillespie commented on sdk/nodejs/resource.ts:
```diff
@@ -269,6 +269,13 @@ export abstract class CustomResource extends Resource {
+     * classes that inherit from `CustomResource`.
+     */
+    // tslint:disable-next-line:variable-name
+    /* @internal */ public readonly __pulumiType: string;
```
This is interesting in its own right - do we need to make tfgen/kubernetes changes to accommodate this?

hausdorff replied:
Yes we do.

swgillespie replied:
👍

hausdorff commented:
> If I'm understanding this right, this invokes the language host with a limited resource monitor that isn't allowed to do resource registrations, right? I would expect this not to work, because the language host always registers a `Stack` root resource and registers outputs on it upon program completion. Are we suppressing the registration of the `Stack` resource somewhere?

lol @swgillespie you turn out to be correct. I forgot that I'd built the SDK the other day, and it had some other stuff in it. Will fix soon.

hausdorff commented:
@swgillespie I'm not sure we want to just ignore the `RegisterResource` request from the stack, because it will just come back and ask to register the resource outputs immediately after. I've gone ahead and added it to the language server protocol because I think it makes sense there.

hausdorff commented on pkg/engine/query.go:
```diff
@@ -0,0 +1,198 @@
+	//      that were loaded are recorded. Second, all first class providers record the version of the plugin
+	//      to which they are bound.
+	//
+	// In order to get a complete view of the set of plugins that we need for an update, we must consult both
```
I thought I did, but I guess I missed one.

lukehoban commented on pkg/engine/query.go:
```diff
@@ -0,0 +1,196 @@
+	//
+	// In order to get a complete view of the set of plugins that we need for an query, we must consult both
+	// sources and merge their results into a list of plugins.
+	languagePlugins, err := gatherPluginsFromProgram(opts.plugctx, plugin.ProgInfo{
```
We're copy/pasting a *lot* of code here.  Can we share some/all of this better? cc @swgillespie for thoughts on structure.

hausdorff replied:
We are. We chose to do it this way rather than to impact the update path. Open to other suggestions, but I actually think of this as a feature, not a bug.

lukehoban replied:
I'll defer to @swgillespie and @ellismg - but it feels like we should extract some of this stuff out and reuse it.  All of the plugin loading here is something that should be usable in both.  It doesn't feel helpful to me to have copies of this.

hausdorff replied:
oh you mean just the plugin stuff (vs the entire PR)? We can probably consolidate that, yes.

lukehoban commented on pkg/resource/deploy/source_query.go:
```diff
@@ -0,0 +1,274 @@
+	// Create a new builtin provider. This provider implements features such as `getStack`.
+	builtins := newBuiltinProvider(client)
+
+	// First, fire up a resource monitor that will watch for and record resource creation.
```
Comment doesn't sound right.

lukehoban commented on sdk/nodejs/cmd/run/run.ts:
```diff
@@ -215,6 +216,28 @@ ${defaultMessage}`);
+                reportModuleLoadFailure(program, e);
+            }
+
+            throw e;
```
Should we extract this out instead of copy/paste?

hausdorff replied:
Yes.

lukehoban commented on pkg/resource/deploy/source_query.go:
```diff
@@ -0,0 +1,274 @@
+				Args:             src.runinfo.Args,
+				Config:           config,
+				DryRun:           true,
+				QueryModeEnabled: true,
```
We don't use the `Enabled` postfix for other booleans.  Should we name this just `QueryMode` or `Query`?

hausdorff replied:
We did, I thought, for test mode.

lukehoban replied:
Not in the same way - only for some internal helpers.  I'd stay consistent with the approach used in the siblings here (and elsewhere this is passed through).

lukehoban commented on sdk/python/lib/pulumi/resource.py:
```diff
@@ -261,6 +267,7 @@ def __init__(self,
                resource.
         """
         Resource.__init__(self, t, name, True, props, opts)
+        self.__pulumiType = t
```
Is this change related to the work for this PR?

hausdorff replied:
I think you could argue no. We can pull it into a separate PR.

lukehoban commented on sdk/python/lib/pulumi/runtime/proto/status_pb2.py:
```diff
@@ -1,3 +1,4 @@
+# -*- coding: utf-8 -*-
```
Are these expected?

hausdorff replied:
I don't know, but our build system runs protoc is inside a docker container so I don't think it's a thing from my machine specifically.

hausdorff replied:
Is there any objection to keeping these? We probably should have them anyway.

lukehoban commented on pkg/backend/display/display.go:
```diff
@@ -37,10 +37,19 @@ func ShowEvents(
-	} else {
+	case DisplayProgress:
 		ShowProgressEvents(op, action, stack, proj, events, done, opts, isPreview)
+	case DisplayQuery:
```
Given that the only place we seem to consume `DisplayQuery` is this location where we Fail - is there any reason it needs to be part of the enum here - vs. some other flag orthogonal to this enum?  It feels like this isn't really another display mode in the same sense as the other two - though I may be misreading some of the implications of this.

hausdorff replied:
I'm not sure I understand the objection, but to me this does seem like it should be alongside diff and progressing view. It's basically a completely separate modality for the display settings, and I actually think that's really important to capture.

lukehoban commented:
Also - I don't see Python support here?

hausdorff commented:
@lukehoban I say in the description I'm working on tests.

hausdorff commented on cmd/query.go:
```diff
@@ -0,0 +1,96 @@
+		Short: "Run query program against cloud resources",
+		Long: "Run query program against cloud resources.\n" +
+			"\n" +
+			"This command loads a Pulumi query project and executes it. Pulumi provides various useful constructs\n" +
```
I don't know which of these is true, but what I'm trying to say is that you need things like package.json. It's not "just" a file.

hausdorff replied:
I changed it, re-open if you still don't like the wording.

hausdorff commented on pkg/backend/backend.go:
```diff
@@ -230,3 +233,34 @@ func (c *backendClient) GetStackOutputs(ctx context.Context, name string) (resou
 	return nil, errors.New("downloading plugins at runtime not available when using local backend")
 }
+
+func (c *backendClient) GetStackResourceOutputs(
```
I thought the style was to comment the interface instead of having like 9 comments that are maybe inconsistent? Can do either way just lmk.

lukehoban replied:
Yeah - that's fine.  Just looking for some place that documents this new interchange type shape.

hausdorff commented on pkg/backend/backend.go:
```diff
@@ -230,3 +233,34 @@ func (c *backendClient) GetStackOutputs(ctx context.Context, name string) (resou
+
+		resc := resource.PropertyMap{
+			resource.PropertyKey("type"):    resource.NewStringProperty(string(r.Type)),
+			resource.PropertyKey("outputs"): resource.NewObjectProperty(r.Outputs)}
```
Reading the code I thought this should be the "clean" version, after everything is taken out. Could be wrong.

hausdorff replied:
cc @ellismg I'll take another look after testing though

hausdorff commented on pkg/backend/filestate/backend.go:
```diff
@@ -487,6 +498,13 @@ func (b *localBackend) apply(
+func (b *localBackend) query(ctx context.Context, stack backend.Stack, op backend.UpdateOperation,
+	events chan<- engine.Event) result.Result {
+
+	panic("not implemented")
```
oh right, I forgot about this, sorry. the PR is big enough that it's starting to get away from me.

hausdorff replied:
As for the second complaint, it's hard to know what you mean specifically here, but the somewhat awkward structure of the code is because the things we need are scattered between the engine and backend packages. We could perform more surgery and pull it out, but I personally don't think it's worth the trouble.

hausdorff replied:
Let's resolve for now, re-open if you have more you want to discuss.

lukehoban commented:
> I say in the description I'm working on tests.

Ahh - indeed - sorry I missed that.  Great.

justinvp commented on CHANGELOG.md:
```diff
@@ -43,6 +43,10 @@
 
+### Improvements
+
+- Add ability to enumerate resource outputs from state snapshot of arbitrary stacks.
```
Looks like this needs to be moved up to be under the unreleased version, v0.17.10 (after rebasing against master).

hausdorff commented:
@lukehoban I wasn't sure we'd have a query model for Python so I left it out. We can totally make a plan, but I think we should do that before exposing query mode through the python sdk, personally.

swgillespie commented:
> @lukehoban I wasn't sure we'd have a query model for Python so I left it out. We can totally make a plan, but I think we should do that before exposing query mode through the python sdk, personally.

In general I think the goal is to keep as close to feature parity as possible. IMO we should have a plan either way before merging this PR to either 1) implement Python support or 2) decide that we're not going to do it for Python, for whatever reason.

swgillespie commented on sdk/nodejs/cmd/run/run.ts:
```diff
@@ -215,6 +216,28 @@ ${defaultMessage}`);
 
     programStarted();
 
+    if (isQueryModeEnabled()) {
```
I think I'd much rather this code live in `stack.ts`, which is where we register the `Stack` resource, e.g.:

```
if (!isQueryModeEnabled()) {
   new Stack(init);
} else {
   init();
}

hausdorff replied:
The test mode stuff is already here, though. What do you think we should do about that?

swgillespie replied:
What do you mean? This file just parses arguments and then goes to `runInPulumiStack`, right? I'd expect the test mode stuff to be really deep in the runtime where we end up doing the actual RPC calls to the engine.

hausdorff replied:
I mean that this is basically identical to the code like 4 lines below -- it won't be after I fix it, though. But I'd planned to keep it all in one file.

hausdorff replied:
Also it would be a little weird to `runInPulumiStack` with no stack, no?

swgillespie replied:
From an API perspective, I think it would be a little weird to run a query without using "runInPulumiStack". In my mind, the name of that function (and the way we've used it) implies that we're running in the context of a stack, which we are even when we are querying. The only difference is that we don't need to register the root Stack and its outputs when we're querying.

I'm fine either way, but I would prefer that we rename or document `runInPulumiStack` aggressively to mean that it should only be used for updates, or we can make `runInPulumiStack` work for queries too. I'd favor the second, but I leave that up to you.

hausdorff replied:
Oh, that makes more sense. Yes let's do it your way.

hausdorff commented:
> In general I think the goal is to keep as close to feature parity as possible. IMO we should have a plan either way before merging this PR to either 1) implement Python support or 2) decide that we're not going to do it for Python, for whatever reason.

The "plan" is that we will do it eventually, but I'm not sure that's what you mean. We currently have no plans to do it in the immediate future, and turning "on" query mode in Python without anything to query seems not great.

swgillespie commented:
> We currently have no plans to do it in the immediate future

That's all I was asking, just that we're clear on that (for the purposes of communication).

hausdorff commented:
Got it, makes sense.

hausdorff commented:
I think I've addressed every change that I know about, but there are a couple open comments from @lukehoban.

@swgillespie @lukehoban @ellismg per offline conversation, I've added a bunch of unit tests for some of the critical components of this work -- but the end-to-end integration tests will need to land with the SDK changes. I've written those, but let's plan for those to get checked in at that point.

@ellismg I think we're not comfortable pushing this in until you sign off. :)

ellismg commented on pkg/backend/backend_test.go:
```diff
@@ -0,0 +1,457 @@
+	// Create a `backendClient` that consults a (mock) `Backend` to make sure it can get the stack
+	// resource outputs correctly.
+
+	resc1 := liveState("liveType1", "resc1", resource.PropertyMap{
```
NIT: I think it would be ideal if the URN And type followed the structure implied by the URN and Type tokens (i.e. they looked more like the real URNs and Types we'd see generated from real programs).

hausdorff replied:
Done.

ellismg commented on pkg/backend/filestate/backend.go:
```diff
@@ -487,6 +498,13 @@ func (b *localBackend) apply(
+func (b *localBackend) query(ctx context.Context, stack backend.Stack, op backend.UpdateOperation,
+	events chan<- engine.Event) result.Result {
+
+	return result.Error("Local backend does not support querying over the state")
```
This is something we will support, right?  Could you mark this with a TODO?  I expect that the implementation will be very close to what we do for http backed stacks?

hausdorff replied:
We can add a TODO. It would be pretty easy to do, I didn't because I wasn't sure we wanted to.

ellismg commented on pkg/engine/query.go:
```diff
@@ -0,0 +1,168 @@
+		return nil, err
+	}
+
+	// Once we've installed all of the plugins we need, make sure that all analyzers and language plugins are
```
I get that this is a copy-pasta'd comment, but we would never load provider plugins in query mode today, right?  (I am guessing we will in the future when we start to query live state...)

ellismg commented on pkg/engine/update.go:
```diff
@@ -131,6 +134,22 @@ func newUpdateSource(
 	}
 
+	// Collect the version information for default providers.
+	defaultProviderVersions := computeDefaultProviderPlugins(languagePlugins, allPlugins)
```
Can you just help me understand what's going on with this refactoring?  I would expect we would not need to make any changes to how the default plugin providers are installed as part of this work, since we should never be loading them?

ellismg replied:
I'm guessing that this just made things simpler, even though we do end up computing default provider plugins in the query case.  Not pushing back on making the code clearer, just trying to understand if it was needed for correctness or not.

hausdorff replied:
Actually we _don't_ end up computing default provider plugins for the query case. At least if I understand your point.

This is not a correctness change -- luke asked me to simplify this so that the plugin loading code is shared between the query and update modes. I had to move this out to the update path so that we didn't compute this for the query case.

ellismg commented on pkg/backend/backend.go:
```diff
@@ -230,3 +233,34 @@ func (c *backendClient) GetStackOutputs(ctx context.Context, name string) (resou
+		resc := resource.PropertyMap{
+			resource.PropertyKey("type"):    resource.NewStringProperty(string(r.Type)),
+			resource.PropertyKey("outputs"): resource.NewObjectProperty(r.Outputs)}
+		pm[resource.PropertyKey(r.URN)] = resource.NewObjectProperty(resc)
```
I do wonder if we should return something richer than a raw PropertyMap here, and have the implementation of `readStackResourceOutputs` in the built in provider translate that into a meaninful property map.  Right now the shape of the final type we are going to send over the wire is split between this method and that one, but I think it would be cleaner if we didn't return raw property maps until much later.  If we just returned a map[URN]struct{tokens.Type, resource.PropertyMap} here, does that make things overall cleaner?

hausdorff replied:
Let's do it this way. I don't think it makes it much clearer, and this way the signature differs from `GetStackOutputs` (which is what this is modeled after) but I also don't think it's bad enough to dissent on.

hausdorff replied:
@lukehoban @ellismg actually, the whole stack ref reading routines are implemented as we implement this new stuff. I'd prefer to keep them consistent. Objections?

ellismg commented on CHANGELOG.md:
```diff
@@ -7,6 +7,8 @@
 - Depend on newer version of gRPC package in the NodeJS SDK. This version has
   prebuilt binaries for Node 12, which should make installing `@pulumi/pulumi`
   more reliable when running on Node 12.
+- Add `pulumi query` command for querying cloud resources. Includes support for querying state
```
Can we just drop this from the changelog for now given that we are hiding the command behind the a debug flag (and then add it back when we actually have the E2E MVP)?

hausdorff replied:
Yes we can.

hausdorff commented on cmd/query.go:
```diff
@@ -0,0 +1,97 @@
+				return result.FromError(err)
+			}
+
+			m, err := getUpdateMetadata("" /*msg*/, root)
```
Yep.

hausdorff commented:
Alright, @lukehoban @ellismg @swgillespie I'm going to ship this. If you want me to follow up on anything else ping me and we can figure it out.
